service/cri.go

```diff
@@ -26,18 +26,34 @@ import (
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
+// TODO: switch to "github.com/containerd/containerd/pkg/snapshotters" once all tools using
+//
+//	stargz-snapshotter (e.g. k3s) move to containerd version where that pkg is available.
 const (
-	// targetRefLabel is a label which contains image reference passed from CRI plugin.
+	// targetRefLabel is a label which contains image reference and will be passed
+	// to snapshotters.
 	targetRefLabel = "containerd.io/snapshot/cri.image-ref"
-
-	// targetDigestLabel is a label which contains layer digest passed from CRI plugin.
-	targetDigestLabel = "containerd.io/snapshot/cri.layer-digest"
-
+	// targetLayerDigestLabel is a label which contains layer digest and will be passed
+	// to snapshotters.
+	targetLayerDigestLabel = "containerd.io/snapshot/cri.layer-digest"
 	// targetImageLayersLabel is a label which contains layer digests contained in
-	// the target image and is passed from CRI plugin.
+	// the target image and will be passed to snapshotters for preparing layers in
+	// parallel. Skipping some layers is allowed and only affects performance.
 	targetImageLayersLabel = "containerd.io/snapshot/cri.image-layers"
 )
 
+const (
+	// targetImageURLsLabelPrefix is a label prefix which constructs a map from the layer index to
+	// urls of the layer descriptor. This isn't contained in the set of the labels passed from CRI plugin but
+	// some clients (e.g. nerdctl) passes this for preserving url field in the OCI descriptor.
+	targetImageURLsLabelPrefix = "containerd.io/snapshot/remote/urls."
```

Comment:
Should nerdctl use the CRI labels too?

Ok to preserve old labels for compatibility sake though

Reply:
nerdctl shoudl use containerd CRI labels (`targetRefLabel`+`targetLayerDigestLabel`+`targetImageLayersLabel`) + extra labels (`targetImageURLsLabelPrefix`+`targetURLsLabel`; they aren't defined/used by CRI). All of them should be needed for nerdctl to perform registry-based lazy pulling as well as IPFS-based lazy pulling.